src/mesh/voxel_cloud.rs

```diff
@@ -448,6 +448,165 @@ impl ScalarField {
+                        Point3::from(corners_cartesian_coordinates[3].coords.lerp(&corners_cartesian_coordinates[7].coords, 0.5)), //e11
+                    ];
+
+                    // Boolean vector marking each of the voxels either volume
+                    // (`true`) or void(`false`)
+                    let corners_inside_volume_pattern: Vec<_> = corners_absolute_coordinates
+                        .iter()
+                        .map(|a| {
+                            self.value_at_absolute_voxel_coordinate(&a)
+                                .map(|value| volume_value_range.contains(&value))
+                                .unwrap_or(false)
+                        })
+                        .collect();
+
+                    // The marching cubes lookup table contains 256 possible
+                    // samples of mesh patches. The sample index specifies the
+                    // proper sample for the current voxel group.
+                    let mut sample_index = 0_u8;
+                    if corners_inside_volume_pattern[0] {
+                        sample_index |= 1;
+                    }
+                    if corners_inside_volume_pattern[1] {
+                        sample_index |= 2;
+                    }
+                    if corners_inside_volume_pattern[2] {
+                        sample_index |= 4;
+                    }
+                    if corners_inside_volume_pattern[3] {
+                        sample_index |= 8;
+                    }
+                    if corners_inside_volume_pattern[4] {
+                        sample_index |= 16;
+                    }
+                    if corners_inside_volume_pattern[5] {
+                        sample_index |= 32;
+                    }
+                    if corners_inside_volume_pattern[6] {
+                        sample_index |= 64;
+                    }
+                    if corners_inside_volume_pattern[7] {
+                        sample_index |= 128;
+                    }
+
+                    // The marching cubes lookup table sample, that contains
+                    // triplets of indices to the edge midpoints (currently
+                    // already vertices) that will become mesh patch faces.
+                    let marching_cubes_faces =
+                        marching_cubes_lookup_table[cast_usize(sample_index)].clone();
+
+                    // In some cases (when all the voxels are inside or outside
+                    // a volume), there is no mesh patch to be created.
+                    if !marching_cubes_faces.is_empty() {
+                        marching_cubes.push(
+                        // Generate new mesh path from all edge midpoint
+                        // vertices and only those faces that should be created
+                        // according th the marching cubes lookup table.
+                        Mesh::from_triangle_faces_with_vertices_and_computed_normals_remove_orphans(
```

Comment:
`remove_orphan_vertices` is kind of wasteful in this situation (called by this mesh constructor). A by hand implementation could be way more efficient. Let's leave an optimization fixme here.

Reply:
This is the optimization I mentioned yesterday. Cleaning up the list of vertices before feeding them into a mesh constructor, is exactly what `remove_orphan_vertices` does. So we either rely on an existing routine or we somehow put a very similar piece of code here. I decided not to do it here.

There is one more option - removing all the orphans all at once when all the mesh patches are joined into a single mesh. This, however, sounds way slower as the orphan removal will have to scan through a much longer list of items.

Am I missing something?

Reply:
> So we either rely on an existing routine or we somehow put a very similar piece of code here. I decided not to do it here.

`remove_orphan_vertices` is general purpose and allocates. We could pre-filter without allocation with the information we have in this function.